xcdat/temporal.py

```diff
@@ -120,40 +127,126 @@ class TemporalAccessor:
 
     def average(
+        self,
+        data_var: str,
+        weighted: bool = True,
+        center_times: bool = False,
+    ):
+        """
+        Returns a Dataset with the average of a data variable and the time
+        dimension removed.
+
+        This method is particularly useful for calculating the weighted averages
+        of monthly or yearly time series data because the number of days per
+        month/year can vary based on the calendar type, which can affect
+        weighting. For unweighted time averages or other frequencies, you can
+        set ``weighted=False``.
+
+        Parameters
+        ----------
+        data_var: str
+            The key of the data variable for calculating averages
+
+        weighted : bool, optional
+            Calculate averages using weights, by default True.
+
+            Weights are calculated by first determining the length of time for
+            each coordinate point using the difference of its upper and lower
+            bounds. The time lengths are grouped, then each time length is
+            divided by the total sum of the time lengths to get the weight of
+            each coordinate point.
+
+        center_times: bool, optional
+            If True, center time coordinates using the midpoint between its
+            upper and lower bounds. Otherwise, use the provided time
+            coordinates by default False.
+
+        Returns
+        -------
+        xr.Dataset
+            Dataset with the average of the data variable and the time dimension
+            removed.
+
+        Examples
+        --------
+
+        Get weighted averages for a monthly time series data variable:
+
+        >>> ds_month = ds.temporal.average("ts", freq="month", center_times=False)
+        >>> ds_month.ts
+        """
+        freq = self._infer_freq()
+
+        return self._averager(data_var, "average", freq, weighted, center_times)
+
+    def _infer_freq(self) -> Frequency:
+        """Infers the time frequency from the coordinates.
+
+        This method infers the time frequency from the coordinates by
+        calculating the minimum delta and comparing it against a set of
+        conditionals.
+
+        The native ``xr.infer_freq()`` method does not work for all cases
+        because the frequency can be irregular (e.g., different hour
+        measurements), which ends up returning None.
+
+        Returns
+        -------
+        Frequency
+            The time frequency.
+        """
+        time_coords = self._dataset[self._dim_name]
+        min_delta = pd.to_timedelta(np.diff(time_coords).min(), unit="ns")
+
+        if min_delta < pd.Timedelta(days=1):
+            return "hour"
+        elif min_delta >= pd.Timedelta(days=1) and min_delta < pd.Timedelta(days=28):
+            return "day"
+        elif min_delta >= pd.Timedelta(days=28) and min_delta < pd.Timedelta(days=365):
+            return "month"
+        else:
+            return "year"
```

Comment:
@lee1043 I added this new method, `_infer_freq()` for inferring the `freq` of the coordinates. It will only be used if `weighed=True`.

Let me know how the algorithm looks, thanks.

Reply:
@tomvothecoder I think this is great and very straightforward, thank you for revising it. I am playing with the new function in the notebook now. 

P.S: (typo in above comment `weighed=True` : `weighed` --> `weighted`